pep-0723.rst

```diff
@@ -0,0 +1,185 @@
+
+This PEP chooses to follow the latest developments of other modern packaging
+ecosystems (namely `Rust <https://github.com/rust-lang/rfcs/blob/master/text/3424-cargo-script.md>`__
+and `Go <https://github.com/erning/gorun>`__) by embedding the existing
+standard :pep:`621` that is used to describe projects.
+
+Any Python script may assign a variable named ``__pyproject__`` to a multi-line
+double-quoted string containing a valid TOML document. This document contains
+what you would expect to see in a ``pyproject.toml`` file, with the following
+exceptions regarding the ``[project]`` table:
+
+* The ``name`` and ``version`` fields MUST be defined dynamically by tools if
+  the user does not define them
+* These fields do not need to be listed in the ``dynamic`` array
+
+Example
+-------
+
+The following is an example of a script with an embedded ``pyproject.toml``:
+
+.. code:: python
+
+    __pyproject__ = """
+    [project]
+    requires-python = ">=3.11"
+    dependencies = [
+      "requests<3",
+      "rich",
+    ]
+    """
+
+    import requests
+    from rich.pretty import pprint
+
+    resp = requests.get("https://peps.python.org/api/peps.json")
+    data = resp.json()
+    pprint([(k, v["title"]) for k, v in data.items()][:10])
+
+
+Reference Implementation
+========================
+
+This regular expression may be used to parse the metadata::
+
+    (?ms)^__pyproject__ *= *"""$(.+?)^"""$
+
+For languages that do not support easily accessing the match group, one may
+parse the entire match as TOML and then access the ``__pyproject__`` key.
+
+The following is an example of how to read the metadata on Python 3.11 or
+higher.
+
+.. code:: python
+
+    import re, tomllib
+
+    def read(script: str) -> dict | None:
+        match = re.search(r'(?ms)^__pyproject__ *= *"""$(.+?)^"""$', script)
+        return tomllib.loads(match.group(1)) if match else None
+
+Often tools will provide commands to manage dependencies. The following is a
+crude example of modifying the content using the ``tomlkit`` library.
+
+.. code:: python
+
+    import re, tomlkit
+
+    def add(script: str, dependency: str) -> str:
+        match = re.search(r'(?ms)^__pyproject__ *= *"""$(.+?)^"""$', script)
+        config = tomlkit.parse(match.group(1))
+        config['project']['dependencies'].append(dependency)
+
+        start, end = match.span(1)
+        return script[:start] + tomlkit.dumps(config) + script[end:]
+
+
+Backwards Compatibility
+=======================
+
+At the time of writing, ``__pyproject__`` only appears 5 times
+`on GitHub <https://github.com/search?q=__pyproject__&type=code>`__ and 4 of
+those belong to a user who appears to already be using this PEP's format
+for its intended purpose.
+
+
+How to Teach This
+=================
+
+The format is intended to bridge the gap between users just writing scripts and
+users writing packages. Knowledge of how to write metadata for one use case will
+be directly transferable to the other.
```

Comment:
The target audience for single-file scripts will include system admins and other groups who will not know TOML. How will such users learn this format? They will also have no existing knowledge of the `pyproject.toml` format, so will have no existing knowledge here. And worse, a lot of the documentation explaining `pyproject.toml` will be discussion concepts that are irrelevant and confusing to them.

Reply:
The PEP has been rewritten and I think now addresses the experience of different types of users such as sysadmins